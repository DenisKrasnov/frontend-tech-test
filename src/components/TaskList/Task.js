import React from "react";
import { shape, string, func } from "prop-types";

import Icon from "../Icon";
import styles from "./Task.scss";

const Task = props => {
  const onEdit = () => {
    const { task, openEditModal } = props;
    openEditModal(task);
  };

  const onDelete = () => {
    const { task, deleteTask, clearError } = props;
    clearError();
    deleteTask({ id: task.id });
  };

  const { task } = props;

  return (
    <div className={styles.taskContainer}>
      <section className={styles.task}>
        <h4 className={styles.title}>{task.title}</h4>
        <p className={styles.description}>{task.description}</p>
      </section>
      <div className={styles.actions}>
        <button className={styles.iconButton} onClick={onDelete} type="button">
          <Icon glyph="trash" />
        </button>
        <button className={styles.iconButton} onClick={onEdit} type="button">
          <Icon glyph="pencil" />
        </button>
      </div>
    </div>
  );
};

Task.propTypes = {
  task: shape({
    id: string.isRequired,
    title: string,
    description: string,
  }).isRequired,
  openEditModal: func.isRequired,
  deleteTask: func.isRequired,
  clearError: func.isRequired,
};

export default Task;
