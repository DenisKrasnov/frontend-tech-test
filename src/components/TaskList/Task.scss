.taskContainer {
  background: #fff;
  box-shadow: 0 1px 5px $shadow-color;
  margin: 10px;
}

.task {
  display: flex;
  flex-direction: column;
  padding: 0 20px;
  width: 300px;
  height: 210px;
}

.title {
  margin-block-end: 10px;
  overflow-wrap: break-word;
}

.description {
  margin: 0;
  overflow: hidden;
  overflow-wrap: break-word;
}

.iconButton {
  display: inline-block;
  padding: 4px;
  border: 0;
  outline: 0;
  margin-left: 10px;

  &:hover {
    svg {
      fill: $icon-hover-color;
    }
  }
}

.actions {
  display: flex;
  flex-direction: row-reverse;
  padding: 10px;
  background-color: #fff;
  border-top: 1px dashed $border-color;
}
