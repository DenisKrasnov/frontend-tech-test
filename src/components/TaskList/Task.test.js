import React from "react";
import { shallow } from "enzyme";

import Task from "./Task";

const noop = () => {};
const task = {
  id: "2",
  title: "__TITLE__",
  description: "__DESCRIPTION__",
};

describe("Task", () => {
  it("renders", () => {
    const component = shallow(
      <Task
        clearError={noop}
        deleteTask={noop}
        openEditModal={noop}
        task={task}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("triggers the 'deleteTask' and 'clearError'", () => {
    const deleteTaskSpy = jest.fn();
    const clearErrorSpy = jest.fn();
    const component = shallow(
      <Task
        clearError={clearErrorSpy}
        deleteTask={deleteTaskSpy}
        openEditModal={noop}
        task={task}
      />
    );

    component
      .find(".actions")
      .childAt(0)
      .simulate("click");
    expect(clearErrorSpy).toHaveBeenCalledTimes(1);
    expect(deleteTaskSpy).toHaveBeenCalledWith({ id: task.id });
  });

  it("tests the state when the 'openEditModal'", () => {
    const openEditModalSpy = jest.fn();
    const component = shallow(
      <Task
        clearError={noop}
        deleteTask={noop}
        openEditModal={openEditModalSpy}
        task={task}
      />
    );

    component
      .find(".actions")
      .childAt(1)
      .simulate("click");
    expect(openEditModalSpy).toHaveBeenCalledWith(task);
  });
});
