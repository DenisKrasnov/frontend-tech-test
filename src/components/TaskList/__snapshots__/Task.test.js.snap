// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task renders 1`] = `
<div
  className="taskContainer"
>
  <section
    className="task"
  >
    <h4
      className="title"
    >
      __TITLE__
    </h4>
    <p
      className="description"
    >
      __DESCRIPTION__
    </p>
  </section>
  <div
    className="actions"
  >
    <button
      className="iconButton"
      onClick={[Function]}
      type="button"
    >
      <Icon
        color="blue"
        glyph="trash"
        onClick={[Function]}
      />
    </button>
    <button
      className="iconButton"
      onClick={[Function]}
      type="button"
    >
      <Icon
        color="blue"
        glyph="pencil"
        onClick={[Function]}
      />
    </button>
  </div>
</div>
`;
